File: 2020/17/Haskell/src/Part2.hs
module Part2 where

import           Data.HashMap.Strict            ( HashMap )
import qualified Data.HashMap.Strict           as HM
import           Data.List

type Point = (Int, Int, Int, Int)
type HyperCube = HashMap Point Char

initHyperCube :: Point -> HyperCube -> String -> HyperCube
initHyperCube _ c [] = c
initHyperCube p c (x : xs)
  | x == '\n' = initHyperCube (px + 1, 0, pz, pw) c xs
  | otherwise = initHyperCube (px, py + 1, pz, pw) c' xs
 where
  (px, py, pz, pw) = p
  c'               = HM.insert p x c

hyperCubeLookup :: Point -> HyperCube -> Char
hyperCubeLookup = HM.findWithDefault '.'

getNeighbours :: Point -> [Point]
getNeighbours (x, y, z, w) =
  [ (a, b, c, d)
  | d <- w'
  , c <- z'
  , b <- y'
  , a <- x'
  , (a, b, c, d) /= (x, y, z, w)
  ]
 where
  f t = [t - 1 .. t + 1]
  (x', y', z', w') = (f x, f y, f z, f w)

getUniqueNeighbours :: [Point] -> [Point]
getUniqueNeighbours = map head . group . sort . concatMap getNeighbours

countActiveNeighbours :: HyperCube -> [Point] -> Int
countActiveNeighbours c = length . filter ((== '#') . (`hyperCubeLookup` c))

stepHyperCube :: HyperCube -> HyperCube -> [Point] -> HyperCube
stepHyperCube _  c []       = c
stepHyperCube c0 c (p : ps) = stepHyperCube c0 c' ps
 where
  this       = hyperCubeLookup p c0
  neighbours = getNeighbours p
  active     = countActiveNeighbours c0 neighbours
  c' | this == '#' && active `notElem` [2, 3] = HM.insert p '.' c
     | this == '.' && active == 3             = HM.insert p '#' c
     | otherwise                              = c

cycleHyperCube :: Int -> Int -> HyperCube -> HyperCube
cycleHyperCube n0 n c | n == n0   = c
                      | otherwise = cycleHyperCube n0 (n + 1) c'
 where
  k  = HM.keys c
  k' = k ++ getUniqueNeighbours k
  c' = stepHyperCube c c k'

part2 :: String -> Int
part2 =
  length
    . filter (== '#')
    . HM.elems
    . cycleHyperCube 6 0
    . initHyperCube (0, 0, 0, 0) HM.empty

File: 2020/17/Rust/src/part_2.rs
use std::collections::{HashMap, HashSet};

type Point = (i32, i32, i32, i32);
type Grid = HashMap<Point, char>;

pub fn init_grid(input: &str) -> Grid {
    let mut x = 0;
    let mut y = 0;
    let z = 0;
    let w = 0;
    let mut grid: Grid = HashMap::new();
    for hyper_cube in input.chars() {
        if hyper_cube == '\n' {
            x += 1;
            y = 0;
        } else {
            grid.insert((x, y, z, w), hyper_cube);
            y += 1;
        }
    }
    grid
}

fn grid_lookup(point: &Point, grid: &Grid) -> char {
    if let Some(c) = grid.get(point) {
        return *c;
    }
    '.'
}

fn get_neighbours(point: &Point) -> HashSet<Point> {
    let (x, y, z, w) = point;
    let neighbouring = |t: &i32| t - 1..=t + 1;
    let mut neighbours = HashSet::new();
    for a in neighbouring(x) {
        for b in neighbouring(y) {
            for c in neighbouring(z) {
                for d in neighbouring(w) {
                    if (&a, &b, &c, &d) != (x, y, z, w) {
                        neighbours.insert((a, b, c, d));
                    }
                }
            }
        }
    }
    neighbours
}

fn count_active_neighbours(grid: &Grid, points: &HashSet<Point>) -> usize {
    points
        .iter()
        .filter(|p| grid_lookup(p, grid) == '#')
        .count()
}

fn step_grid(grid: &Grid, points: &HashSet<Point>) -> Grid {
    let mut next_grid = grid.to_owned();
    for point in points {
        let hyper_cube = grid_lookup(point, grid);
        let neighbours = get_neighbours(point);
        let active = count_active_neighbours(grid, &neighbours);
        if (hyper_cube == '#') && !(2..=3).contains(&active) {
            next_grid.insert(*point, '.');
        } else if (hyper_cube == '.') && (active == 3) {
            next_grid.insert(*point, '#');
        }
    }
    next_grid
}

pub fn cycle_grid(grid: &Grid) -> usize {
    let mut next_grid = grid.to_owned();
    for _ in 0..6 {
        let mut next_keys = next_grid.keys().cloned().collect::<HashSet<_>>();
        for key in next_grid.keys() {
            for next_key in get_neighbours(key) {
                next_keys.insert(next_key);
            }
        }
        next_grid = step_grid(&next_grid, &next_keys);
    }
    next_grid.values().filter(|c| c == &&'#').count()
}
